### PR TITLE
Feature/load image

### PR DESCRIPTION
**Pull request by ChrisBerg**
Added the ability to upload an image and set it as the background image of the map of the current project of the active user.

### CODE REVIEW

File: src/main/java/nl/tudelft/thefirstorder/domain/Map.java [hunk 1]
```diff
@@ -71,6 +74,22 @@ public String getName() {
     }
 
     /**
+     * Returns the location of the background image in the form of n url.
```
Comment (trombomart) [inline src/main/java/nl/tudelft/thefirstorder/domain/Map.java]: Spelling mistake: An url

File: src/main/java/nl/tudelft/thefirstorder/domain/Map.java [hunk 2]
```diff
@@ -71,6 +74,22 @@ public String getName() {
...
+     * @return an url pointing to the location of the background image.
+     */
+    public String getBackground_image() {
+        return background_image;
+    }
+
+    /**
+     * Sets the location of the background image to the given url.
+     * @param background_image should be a url representing the location of an image
```
Comment (trombomart) [inline src/main/java/nl/tudelft/thefirstorder/domain/Map.java]: Spelling mistake: an url

File: src/main/java/nl/tudelft/thefirstorder/service/MapService.java [hunk 3]
```diff
@@ -13,38 +13,38 @@
...
      *  @param pageable the pagination information
      *  @return the list of entities
      */
     Page<Map> findAll(Pageable pageable);
+    /**
+     *  Get all the maps where Project is null.
+     *  
+     *  @return the list of entities
+     */
+    List<Map> findAllWhereProjectIsNull();
```
Comment (trombomart) [inline src/main/java/nl/tudelft/thefirstorder/service/MapService.java]: Why did you add this method for image upload?

Comment (trombomart) [inline src/main/java/nl/tudelft/thefirstorder/service/MapService.java]: Never mind: you switched only the position :p

File: src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java [hunk 4]
```diff
@@ -1,97 +1,91 @@
...
-
+    
     /**
      * Save a map.
-     *
+     * 
      * @param map the entity to save
      * @return the persisted entity
      */
-    @Transactional
```
Comment (trombomart) [inline src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java]: Why not transactional anymore?

File: src/main/java/nl/tudelft/thefirstorder/domain/Map.java [hunk 5]
```diff
@@ -34,6 +34,9 @@
     @Column(name = "name")
     private String name;
 
+    @Column(name = "background_image")
+    private String background_image;
```
Comment (hungnguyen94) [inline src/main/java/nl/tudelft/thefirstorder/domain/Map.java]: You should use camelcase for Java variables. Using underscores isn't very pretty. `backgroundImage` instead of `background_image`.

File: src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java [hunk 6]
```diff
@@ -15,12 +15,7 @@
 import org.springframework.http.MediaType;
 import org.springframework.http.ResponseEntity;
 import org.springframework.transaction.annotation.Transactional;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RequestMethod;
-import org.springframework.web.bind.annotation.RequestParam;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
Comment (trombomart) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java]: Don't use star import

File: src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java [hunk 7]
```diff
@@ -36,16 +31,15 @@
...
     @Inject
     private MapService mapService;
 
     /**
      * POST  /maps : Create a new map.
      *
      * @param map the map to create
-     * @return the ResponseEntity with status 201 (Created) and with body the new map,
-     *      or with status 400 (Bad Request) if the map has already an ID
+     * @return the ResponseEntity with status 201 (Created) and with body the new map, or with status 400 (Bad Request) if the map has already an ID
```
Comment (trombomart) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java]: This line has too many characters on one line

File: src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java [hunk 8]
```diff
@@ -55,9 +49,7 @@
     public ResponseEntity<Map> createMap(@RequestBody Map map) throws URISyntaxException {
         log.debug("REST request to save Map : {}", map);
         if (map.getId() != null) {
-            return ResponseEntity.badRequest().headers(
-                    HeaderUtil.createFailureAlert("map", "idexists", "A new map cannot already have an ID")
-            ).body(null);
+            return ResponseEntity.badRequest().headers(HeaderUtil.createFailureAlert("map", "idexists", "A new map cannot already have an ID")).body(null);
```
Comment (trombomart) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java]: This line has also too many characters on one line

File: src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java [hunk 9]
```diff
@@ -1,97 +1,91 @@
...
+        Map map = mapRepository.findOne(id);
+        return map;
     }
 
     /**
      *  Delete the  map by id.
      *  
      *  @param id the id of the entity
      */
-    @Transactional
```
Comment (hungnguyen94) [inline src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java]: You shouldn't remove this annotation.

File: src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.*;
```
Comment (trombomart) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java]: Don't use star import

File: src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+        return ResponseEntity.ok()
+            .headers(HeaderUtil.createAlert("Uploaded", "image"))
+            .body("{\"location\" : \"" + res + "\"}");
+    }
+
+    private static String writeToFile(InputStream contents) {
+        Random random = new Random();
+        String root = "src/main/webapp/";
+        String location = "content/upload/" + random.nextInt(1000000000) + ".jpg";
```
Comment (trombomart) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java]: Is this really an optimal solution?

File: src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java [hunk 12]
```diff
@@ -1,97 +1,91 @@
...
         mapRepository.delete(id);
     }
-
-    /**
-     * Get all maps where project is null.
-     *
-     * @return List of maps.
-     */
-    @Override
-    @Transactional(readOnly = true)
```
Comment (hungnguyen94) [inline src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java]: This was removed, but another method was added which does the same. The other method is longer though. I'd keep this one, and remove the method you added.

File: src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java (hunk 9 above)
Comment (ChrisBerg) [inline src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java]: Got deleted by accident during addition of the background image field. Reintroducing it.

File: src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java (hunk 6 above)
Comment (hungnguyen94) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java]: Don't use star imports. Checkstyle should warn you for this.

File: src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java (hunk 7 above)
Comment (hungnguyen94) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java]: Checkstyle indentation warning.

File: src/main/webapp/app/map/map-table-view-template.html [hunk 13]
```diff
@@ -22,7 +29,7 @@
         </td>
     </tr>
     <tr class="info" ng-repeat-end>
-    <!--<tr ng-repeat-end uib-collapse="false">-->
+        <!--<tr ng-repeat-end uib-collapse="false">-->
```
Comment (trombomart) [inline src/main/webapp/app/map/map-table-view-template.html]: Comment can be removed

File: src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java [hunk 14]
```diff
@@ -70,8 +62,8 @@
      *
      * @param map the map to update
      * @return the ResponseEntity with status 200 (OK) and with body the updated map,
-     *      or with status 400 (Bad Request) if the map is not valid,
-     *      or with status 500 (Internal Server Error) if the map couldnt be updated
+     * or with status 400 (Bad Request) if the map is not valid,
```
Comment (hungnguyen94) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java]: Checkstyle indentation warning

File: src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java (hunk 10 above)
Comment (hungnguyen94) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java]: Remove star import.

File: src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java (hunk 11 above)
Comment (hungnguyen94) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java]: Random integer?

File: src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+
+        String path = root + location;
+
+        try {
+            int read = 0;
+            byte[] bytes = new byte[1024];
```
Comment (hungnguyen94) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java]: Hardcoded magic number

File: src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java (hunk 12 above)
Comment (ChrisBerg) [inline src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java]: True that, reverted it.

File: src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java (hunk 6 above)
Comment (ChrisBerg) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java]: True, removing the star import

File: src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        String res = writeToFile(file.getInputStream());
+
+        return ResponseEntity.ok()
+            .headers(HeaderUtil.createAlert("Uploaded", "image"))
+            .body("{\"location\" : \"" + res + "\"}");
+    }
+
+    private static String writeToFile(InputStream contents) {
+        Random random = new Random();
+        String root = "src/main/webapp/";
```
Comment (hungnguyen94) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java]: Hardcoded path

File: src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java (hunk 7 above)
Comment (ChrisBerg) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java]: splitted it up

File: src/main/resources/config/application.yml [hunk 17]
```diff
@@ -67,3 +67,12 @@ jhipster:
...
         license:
         licenseUrl:
+
+# ----------------------------------------
+# WEB PROPERTIES
+# ----------------------------------------
+
+# MULTIPART (MultipartProperties)
+multipart:
+    max-file-size: 25Mb # Max file size. Values can use the suffixed "MB" or "KB" to indicate a Megabyte or Kilobyte size.
```
Comment (hungnguyen94) [inline src/main/resources/config/application.yml]: 25Mb is 3.125MB
You should probably change this to `25MB`

File: src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java (hunk 14 above)
Comment (ChrisBerg) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/MapResource.java]: Added indentation

File: src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java (hunk 10 above)
Comment (ChrisBerg) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java]: Replaced with single class imports

File: src/main/webapp/app/entities/map/maps.html [hunk 18]
```diff
@@ -4,9 +4,7 @@
     <div class="container-fluid">
         <div class="row">
             <div class="col-xs-4 no-padding-left">
-                <button class="btn btn-primary"
-                        has-any-authority="ROLE_ADMIN,ROLE_DIRECTOR,ROLE_CAMERA_OPERATOR"
```
Comment (hungnguyen94) [inline src/main/webapp/app/entities/map/maps.html]: The authorities should probably not be removed.

**Comment by hungnguyen94** [comment]
A few remarks. The filenames are hardcoded, which isn't an ideal solution. That could be fixed in either another pull request, or this one. 
Other than that, good job!

File: src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java (hunk 11 above)
Comment (ChrisBerg) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java]: I think so, this prevents overwriting images of other projects. Improved this by changing the integer to a random alphanumeric string.

File: src/main/webapp/app/map/map-table-view-template.html (hunk 13 above)
Comment (ChrisBerg) [inline src/main/webapp/app/map/map-table-view-template.html]: Done

File: src/main/java/nl/tudelft/thefirstorder/domain/Map.java (hunk 1 above)
Comment (ChrisBerg) [inline src/main/java/nl/tudelft/thefirstorder/domain/Map.java]: thanks

File: src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java (hunk 15 above)
Comment (ChrisBerg) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java]: Replaced with a randomly generated alphanumeric string

File: src/main/resources/config/application.yml (hunk 17 above)
Comment (ChrisBerg) [inline src/main/resources/config/application.yml]: 25Mb is in fact 25MB in this case. I will change it for clarity.

File: src/main/webapp/app/entities/map/maps.html (hunk 18 above)
Comment (ChrisBerg) [inline src/main/webapp/app/entities/map/maps.html]: True, reintroduced them

**Comment by ChrisBerg** [comment]
Thanks! I will create an issue for the hard coded filenames, so we can fix that later on.

File: src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java (hunk 15 above)
Comment (ChrisBerg) [inline src/main/java/nl/tudelft/thefirstorder/web/rest/UploadResource.java]: Created a field for those numbers.

**Comment by rubenwiersma** [comment]
Nice work!
